### ISSUE TITLE
Allow specifying vault when fetching 1password secrets

### ISSUE THREAD

**Issue by zb140** [twpayne/chezmoi#763]
## Is your feature request related to a problem? Please describe.

Fetching secrets from 1password is frustratingly slow. For example, on my system, each call to `op get` takes around 1800ms.  For my repo with 4 secret fetches, this means `chezmoi apply` takes around 5 full seconds to complete ☹️ 

## Describe the solution you'd like

I recently learned that specifying which vault to fetch a secret item from (using the `--vault` argument) significantly speeds up access -- calls to `op get item --vault` only take around 800ms.  I put together a prototype (over here: zb140/chezmoi@0cf25b57608566f0a2fa2b7f024b7214ce5a51cc), and with the same repo, `chezmoi apply` now completes in about 2.5 seconds.  Still quite a bit slower than I'd like, but a definite improvement.

## Describe alternatives you've considered

I don't think there's a whole lot `chezmoi` itself can do to help with performance of the secret commands -- in my situation, for example, nearly all of the slowness is directly attributable to the `op get` call.

One idea I did have was that perhaps there could be an option to `chezmoi apply` that would skip processing templates with secrets in them. Since my secrets don't change terribly often, as a user I generally know whether it's safe to skip re-fetching them.  But that sounds like it would be hard to implement.

Or perhaps it could be a property of some templates that they're only re-evaluated sometimes, but that isn't great either -- it seems like a big potential source of confusion for users.

## Additional context

Once there's a clear path forward, I'll fix up my prototype as needed and submit a PR.

**Comment by twpayne** [comment]
> Fetching secrets from 1password is frustratingly slow.

Oof. Sad face.

A few thoughts:

> One idea I did have was that perhaps there could be an option to `chezmoi apply` that would skip processing templates with secrets in them. Since my secrets don't change terribly often, as a user I generally know whether it's safe to skip re-fetching them. But that sounds like it would be hard to implement.

Skipping processing templates with secrets in them would be tricky to implement indeed. However, chezmoi already maintains a persistent state to remember which `run_once_` scripts have been run (and soon for the `chezmoi status` command to remember what the last `chezmoi apply` wrote). We could extend this persistent state to optionally cache 1password secrets too. As all of the secrets already exist in plaintext in the user's home directory, I don't think there are any adverse security consequences to this. A command like `chezmoi secret onepassword flush-cache` could then flush the cache so chezmoi re-requests the new secrets from 1password.

Using `--vault` clearly helps. I had a quick look at your branch and would suggest that the `vault` argument can be added as an optional argument to the existing `onepassword` and `onepasswordDocument` functions without breaking backwards compatibility instead of adding the `*WithVault` variants. This is worth doing independently of any cache.

Finally, for version 2 I've been mulling over making `chezmoi apply` concurrent, which would have the effect of parallelizing fetching secrets from 1password across multiple files (i.e. calls to `onepassword` in different files would be concurrent, but multiple calls to `onepassword` in the same file would remain serialized), which would also add a speed-up. Something like [`singleflight`](https://pkg.go.dev/golang.org/x/sync/singleflight?tab=doc) will help.

Realistically though, concurrent `chezmoi apply` is more likely to land around 2.2, and the 1password secret cache will have a much more positive impact on performance in your case anyhow. So I suggest the following sequence of actions:

1. In the short term, add an optional `vault` argument to the `onepassword` and `onepasswordDocument` template functions (PR from you), targeting chezmoi 1.x.
2. In the medium term, add an optional 1password secret cache, targeting chezmoi 2.1.
3. In the long term, make `chezmoi apply` concurrent, targeting chezmoi 2.2 or later.

**Comment by zb140** [comment]
> So I suggest the following sequence of actions:
> 1. In the short term, add an optional vault argument to the onepassword and onepasswordDocument template functions (PR from you), targeting chezmoi 1.x.
> 1. In the medium term, add an optional 1password secret cache, targeting chezmoi 2.1.
> 1. In the long term, make chezmoi apply concurrent, targeting chezmoi 2.2 or later.

Sounds like a great plan to me.  I'll rework my change to use the optional argument approach.  Actually, that was my original plan, but the only way I could find to make an argument optional was to take a variable number of args and then sanity-check at runtime. I wasn't a huge fan of that approach but I don't feel strongly about it.  Would that be how you would do it, or is there another mechanism I haven't thought of?

**Comment by twpayne** [comment]
> Would that be how you would do it, or is there another mechanism I haven't thought of?

I think that's the only way to do it at the moment.

**Comment by twpayne** [comment]
Here's a really ugly hack for solving the `op` speed problem while waiting for the onepassword cache: cache the secrets in the chezmoi config file.

For example, create

`~/.local/share/chezmoi/.chezmoi.toml.tmpl`
```
[data]
    mycachedsecret = "{{ onepassword "uuid" "vault-uuid" }}"
```

This will retrieve the secret when you run `chezmoi init <repo>` and store them in plain text in your config file. You can then refer to it as `{{ .mycachedsecret }}` in your templates.

To re-generate the config file you can run:

```
chezmoi execute-template --init --output=~/.config/chezmoi/chezmoi.toml < ~/.local/share/chezmoi/.chezmoi.toml.tmpl
```

Note that when you run `chezmoi execute-template` you'll have existing to template data defined in the `chezmoi.toml`, so you might need some template logic to avoid using `promptString` if you already have the value you need.

This is horrible, I know.

**Comment by twpayne** [comment]
Another hack, but maybe with some mileage: write a get-secrets script that retrieves your secrets from 1password and then stores them in a local, fast password manager that can read from a local file like `pass`, `gopass`, or KeePassXC, and then retrieve the secrets from the local file in your templates. To refresh your secrets, re-run your get-secrets script.

Advantages:
* Fast
* Secrets managed by appropriate software
* Can be implemented without waiting for chezmoi

Disadvantages:
* Extra complexity

**Comment by zb140** [comment]
This is a great idea; thanks for the suggestion!  It seems like writing a sufficient `get-secrets` script should be pretty straightforward.  As you point out, the extra complexity is a bit of a bummer, but as long as it's automated it shouldn't be too much of a headache...

**Comment by twpayne** [comment]
I think this is sufficiently fixed by @zb140 in #764 and a bit of extra scripting. Please re-open if needed.
